gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java

```diff
@@ -153,12 +124,105 @@ static GoogleCloudStorageGrpcReadChannel open(
+
+    ByteString footerContent = getFooterContent(resourceId, readOptions, stub, footerOffsetInBytes);
+
+    return new GoogleCloudStorageGrpcReadChannel(
+        stub,
+        stubProvider,
+        resourceId,
+        storageObject.getGeneration(),
+        storageObject.getSize(),
+        footerOffsetInBytes,
+        footerContent,
+        readOptions,
+        backOffFactory);
+  }
+
+  private static com.google.google.storage.v1.Object getObjectMetadata(StorageResourceId resourceId,
+      GoogleCloudStorageReadOptions readOptions, StorageBlockingStub stub) throws IOException {
+    try {
+      // TODO(b/151184800): Implement per-message timeout, in addition to stream timeout.
+      GetObjectRequest getObjectRequest = GetObjectRequest.newBuilder()
+          .setBucket(resourceId.getBucketName())
+          .setObject(resourceId.getObjectName())
+          .build();
+      com.google.google.storage.v1.Object storageObject = stub
+          .withDeadlineAfter(readOptions.getGrpcReadMetadataTimeoutMillis(), MILLISECONDS)
+          .getObject(getObjectRequest);
+      // The non-gRPC read channel has special support for gzip. This channel doesn't
+      // decompress gzip-encoded objects on the fly, so best to fail fast rather than return
+      // gibberish unexpectedly.
+      if (storageObject.getContentEncoding().contains("gzip")) {
+        throw new IOException(
+            "Can't read GZIP encoded files - content encoding support is disabled.");
+      }
+      logger.atFiner().log(
+          "Initialized metadata (size=%s, generation=%s) for '%s'",
+          storageObject.getSize(), storageObject.getGeneration(), resourceId);
+      return storageObject;
+    } catch (StatusRuntimeException e) {
+      throw convertError(e, resourceId);
+    }
+  }
+
+  private static ByteString getFooterContent(StorageResourceId resourceId,
+      GoogleCloudStorageReadOptions readOptions, StorageBlockingStub stub, long footerOffset)
+      throws IOException {
+    try {
+      Iterator<GetObjectMediaResponse> footerContentResponse = stub
+          .withDeadlineAfter(readOptions.getGrpcReadTimeoutMillis(), MILLISECONDS)
+          .getObjectMedia(GetObjectMediaRequest.newBuilder()
+              .setReadOffset(footerOffset)
+              .setBucket(resourceId.getBucketName())
+              .setObject(resourceId.getObjectName())
+              .build());
+
+      ByteString footerContent = null;
+      while (footerContentResponse.hasNext()) {
+        GetObjectMediaResponse objectMediaResponse = footerContentResponse.next();
+        if (objectMediaResponse.hasChecksummedData()) {
+          ByteString content = objectMediaResponse.getChecksummedData().getContent();
+          if (footerContent == null) {
+            footerContent = content;
+          } else {
+            footerContent = footerContent.concat(content);
+          }
+        }
+      }
+      if (footerContent == null) {
```

Comment:
I wonder when `footerContent` could be `null`? Only when `content` is, isn't it? If so, why is this log not printed out earlier?

Reply:
`footerContent` shoudn't be null if `open` succeeded, In case of any modifications in the future, it would be hard to determine and verify this usage. Added this just condition as a defensive approach